README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
simulateR
=========

The `simulateR` R package includes a set of functions used to generate and test simulated data. It allows you to run simulations while easily varying parameters. One common use case would be to estimate statistical power for complex models.

Please be aware that this package is still in development, and as such, bugs may still exist, and functions and function parameters may still be subject to change.

To install the most recent stable release, use the following code:

``` r
install.packages("devtools")
devtools::install_github("jeff-hughes/simulateR@v0.1.3", build_vignettes=TRUE)
```

### Installation Issues

Networked computers can sometimes result in installation issues, as the `install_github` function sometimes has difficulty with networked directories. If this happens to you, use the `.libPaths()` function to find the path to your R libraries. That will likely give you a path starting with two backslashes, but you will need to convert that to a path starting with a drive letter (e.g., 'C:', 'D:'). From there, use the following code:

``` r
install.packages("devtools")
devtools::install_github("jeff-hughes/simulateR@v0.1.3", build_vignettes=TRUE,
    args=c('--library="N:/path/to/libraries/"'))
```

Obviously, change the path to the path where your R libraries are stored.


NEWS.md
<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->
simulateR 0.1.3
===============

MAJOR CHANGES

-   Added `output` parameter that controls how simulation output is provided. This change offers greater flexibility over how the provided function sends back its own output for a single simulation. By default, `simulate()` will now output a list with the data for each simulation; however, if output can be coerced to a data frame or vector, these are offered as options as well.

NEW FEATURES

-   Timing information for how long the simulations took is now provided.

simulateR 0.1.2
===============

NEW FEATURES

-   Added support for playing a sound once simulations finish.

simulateR 0.1.1
===============

NEW FEATURES

-   Added vignette about simulating statistical power.

simulateR 0.1.0
===============

NEW FEATURES

-   `simulate` function that runs simulations.

-   `gen_data` function for generating simulated data from a factor matrix and effects matrix.


R/helpers.R
#' Return results of a simulation.
#'
#' \code{results} is a generic function that extracts the raw data from a
#' simulation.
#'
#' @param sim A simulation object.
#' @param ... Additional arguments to be passed to the particular method for the
#'   given object tye.
#' @return The form of the value returned by \code{results} depends on the class
#'   of its argument. See the documentation of the particular methods for
#'   details of what is produced by that method.
#' @seealso \code{\link{results.simulation}}
#' @export
results <- function(sim, ...) UseMethod('results')


#' Return results of a simulation.
#'
#' \code{results.simulation} returns the raw data from a simulation.
#'
#' @param sim A simulation object of type 'simulation'.
#' @return Returns a data frame with all the data returned from each simulation.
#' @export
results.simulation <- function(sim) {
    return(sim$results)
}


#' Return the tests performed by a simulation.
#'
#' \code{tests} is a generic function that extracts information about the set of
#' specific tests (parameter values) for a simulation.
#'
#' @param sim A simulation object.
#' @param ... Additional arguments to be passed to the particular method for the
#'   given object tye.
#' @return The form of the value returned by \code{tests} depends on the class
#'   of its argument. See the documentation of the particular methods for
#'   details of what is produced by that method.
#' @seealso \code{\link{tests.simulation}}
#' @export
tests <- function(sim, ...) UseMethod('tests')


#' Return the tests performed by a simulation.
#'
#' \code{tests.simulation} extracts information about the set of specific tests
#' (parameter values) for a simulation.
#'
#' @param sim A simulation object of type 'simulation'.
#' @return Returns a data frame with one row for each set of simulations that
#'   was performed.
#' @export
tests.simulation <- function(sim) {
    return(sim$tests)
}


#' Return the number of simulations performed by a simulation.
#'
#' \code{n.sims} is a generic function that extracts information about the
#' number of simulations (per specific test) performed by a simulation.
#'
#' @param sim A simulation object.
#' @param ... Additional arguments to be passed to the particular method for the
#'   given object tye.
#' @return The form of the value returned by \code{n.sims} depends on the class
#'   of its argument. See the documentation of the particular methods for
#'   details of what is produced by that method.
#' @seealso \code{\link{n.sims.simulation}}
#' @export
n.sims <- function(sim, ...) UseMethod('n.sims')


#' Return the number of simulations performed by a simulation.
#'
#' \code{n.sims.simulation} extracts information about the number of simulations
#' (per specific test) performed by a simulation.
#'
#' @param sim A simulation object of type 'simulation'.
#' @return Returns the number of simulations done in each test.
#' @export
n.sims.simulation <- function(sim) {
    return(sim$n.sims)
}


#' Return the timing information of a simulation.
#'
#' \code{timing} is a generic function that extracts the raw data from a
#' simulation.
#'
#' @param sim A simulation object.
#' @param ... Additional arguments to be passed to the particular method for the
#'   given object tye.
#' @return The form of the value returned by \code{timing} depends on the class
#'   of its argument. See the documentation of the particular methods for
#'   details of what is produced by that method.
#' @seealso \code{\link{timing.simulation}}
#' @export
timing <- function(sim, ...) UseMethod('timing')


#' Return the timing information of a simulation.
#'
#' \code{timing.simulation} returns the raw data from a simulation.
#'
#' @param sim A simulation object of type 'simulation'.
#' @return Returns an object of class "proc_time" with information about how
#'   long the simulation process took.
#' @export
timing.simulation <- function(sim) {
    return(sim$timing)
}


#' Calculate error variance given model coefficients.
#'
#' \code{lm_error_var} will calculate the required error variance for a linear
#' model, given specified model coefficients, to create variance for your
#' dependent variable of approximately 'var'.
#'
#' \strong{Note:} This function assumes that \emph{all predictors are
#' independent} (i.e., uncorrelated).
#'
#' @param var The variance you wish your dependent variable to be.
#' @param ... Pass along all model coefficients, excluding the intercept. These
#'   can be named or unnamed.
#' @return Returns the required error variance so that the variance of your
#'   dependent variable is approximately 'var'.
#' @examples
#' lm_error_var(var=1, .15, .3)  # returns error variance of 0.8875
#' @export
lm_error_var <- function(var=1, ...) {
    dots <- list(...)
    exp_var <- 0
    for (i in 1:length(dots)) {
        exp_var <- exp_var + dots[[i]]^2
    }
    return(var - exp_var)
}







R/pwr_sim.R
#' Simulate statistical power.
#'
#' \code{pwr_sim} simulates data by generating samples according to a user-
#' defined function. These data can then be used to examine statistical power.
#' This function is intentionally kept general and flexible, to allow for a wide
#' variety of statistical models.
#'
#' @param func A user-defined function that generates data and performs a
#'   statistical test on that data.
#' @param params A list of parameters to be passed to \code{func}. Each of these
#'   parameters can be a vector of multiple values, and \code{pwr_sim} will test
#'   each combination of these values.
#' @param n.sims Number of simulations (per combination of params).
#' @param boot Whether or not to use bootstrapped data to pass along to your
#'   simulation.
#' @param bootParams If \code{boot=TRUE}, then use \code{bootParams} to pass
#'   along a named list of arguments to the \code{\link{boot}} function. The
#'   statistic and R parameters will be filled automatically, but at minimum you
#'   will need to pass along data. Information about parallel processing will
#'   also be passed along automatically.
#' @param parallel The type of parallel operation to be used (if any).
#' @param ncpus Integer: the number of processes to be used in parallel
#'   operation.
#' @param cl An optional \code{parallel} or \code{snow} cluster for use if
#'   \code{parallel = 'snow'}. If not supplied, a cluster on the local machine
#'   is created for the duration of the simulations.
#' @param ... Additional arguments to be passed to \code{func}. If you do not
#'   need to vary certain parameters in your model, you can pass them to
#'   \code{func} here.
#' @return Returns a data frame with one row per simulation. \code{func} must
#'   return a (named) vector with the results you wish to capture; these vectors
#'   will then be combined for the final output.
#' @seealso \code{\link{boot}}
#' @examples
#' lm_test <- function(N, b0, b1) {
#' x <- rnorm(N, 0, 1)
#' y <- rnorm(N, b0 + b1*x, sqrt(1 - b1^2))
#' data <- data.frame(y, x)
#' model <- lm(y ~ x, data)
#'
#' # capture output from model summary
#' est <- coef(summary(model))['x', 'Estimate']
#' se <- coef(summary(model))['x', 'Std. Error']
#' p <- coef(summary(model))['x', 'Pr(>|t|)']
#'
#' return(c(xm=mean(x), xsd=sd(x), ym=mean(y), ysd=sd(y), est=est, se=se, p=p,
#'     sig=est > 0 & p <= .05))
#' }
#'
#' # test power for two sample sizes: N=200 and N=300, with 5000 sims for each
#' power <- pwr_sim(lm_test, params=list(N=c(200, 300)), n.sims=5000, b0=0, b1=.15)
#' @export
pwr_sim <- function(func, params=NULL, n.sims=5000, boot=FALSE, bootParams=NULL,
    parallel=c('no', 'multicore', 'snow'), ncpus=1, cl=NULL, ...) {

    # cross each param value with every other one, to create all combinations
    if (!is.null(params)) {
        grid <- expand.grid(params, KEEP.OUT.ATTRS=FALSE)
        grid_output <- grid
        names(grid_output) <- paste0(names(grid_output), '.test')
        nSets <- nrow(grid)
    } else {
        grid <- data.frame()  # empty
        grid_output <- data.frame()
        nSets <- 1
    }

    # figure out which parallel method to use
    parallel <- match.arg(parallel)
    have_mc <- FALSE
    have_snow <- FALSE
    if (parallel != 'no' && ncpus > 1) {
        if (requireNamespace('parallel', quietly=TRUE)) {
            if (parallel == 'multicore') {
                have_mc <- .Platform$OS.type != 'windows'
                if (!have_mc) {
                    warning('Multicore is not supported on Windows. Try parallel = "snow" instead. Proceeding with simulations serially.')
                }
            } else if (parallel == 'snow') {
                have_snow <- TRUE
            }
        } else {
            warning("Loading package 'parallel' failed. Proceeding with simulations serially.")
        }
        if (!have_mc && !have_snow) {
            ncpus <- 1
        }
    }

    # set up cluster if necessary
    if (ncpus > 1 && have_snow) {
        if (is.null(cl)) {
            clust <- parallel::makePSOCKcluster(rep('localhost',
                ncpus))
            if (RNGkind()[1] == "L'Ecuyer-CMRG") {
                parallel::clusterSetRNGStream(clust)
            }
        } else {
            clust <- cl
        }
    }

    cat(paste0('Running ',
        prettyNum(nSets * n.sims, big.mark=',', scientific=FALSE),
        ' simulations...\n'))

    allResults <- NULL  # variable to fill with final output
    for (set in 1:nSets) {
        # bootstrap data
        if (boot && 'data' %in% names(bootParams)) {
            boot_output <- do.call(boot::boot, args=c(list(statistic=func,
                R=n.sims, parallel=parallel, ncpus=ncpus, cl=cl), bootParams,
                as.list(grid[set, , drop=FALSE]), ...))

            col_names <- names(boot_output$t0)
            output <- boot_output$t
            colnames(output) <- col_names

        # simulate data
        } else {
            if (ncpus > 1 &&
                (have_mc || have_snow)) {

                if (have_mc) {
                    output <- do.call(parallel::mclapply,
                        args=c(list(X=1:n.sims, FUN=func, mc.cores=ncpus),
                            as.list(grid[set, , drop=FALSE]), ...))
                } else if (have_snow) {
                    output <- do.call(parallel::parLapply,
                        args=c(list(cl=clust, X=1:n.sims, fun=func),
                            as.list(grid[set, , drop=FALSE]), ...))
                }
            } else {
                output <- do.call(lapply, args=c(list(X=1:n.sims, FUN=func),
                    as.list(grid[set, , drop=FALSE]), ...))
            }

            # convert list to data frame
            col_names <- names(output[[1]])
            output <- do.call(rbind.data.frame, output)
            colnames(output) <- col_names
        }

        if (!is.null(params)) {
            extendGrid <- matrix(rep(unlist(grid_output[set, , drop=FALSE]),
                each=n.sims), nrow=n.sims)
            colnames(extendGrid) <- names(grid_output)
            result <- cbind(sim=1:n.sims, extendGrid, output)
        } else {
            result <- cbind(sim=1:n.sims, output)
        }

        if (is.null(allResults)) {
            allResults <- result
        } else {
            allResults <- rbind(allResults, result)
        }
        row.names(allResults) <- NULL
    }

    # stop cluster if we created it
    if (ncpus > 1 && have_snow && is.null(cl)) {
        parallel::stopCluster(clust)
    }

    allResults <- as.data.frame(allResults)

    output <- list(results=allResults, tests=grid, n.sims=n.sims)
    class(output) <- 'pwr_sim'
    return(output)
}






R/summary.R
#' Print summary of simulations.
#'
#' \code{summary.simulation} calculates a summary of a given variable across all,
#' or a portion of, the simulations.
#'
#' @param sim A list of class 'simulation' generated by the function
#'   \code{simulate}.
#' @param outcome A string with the variable name across which to generate the
#'   summaries.
#' @param params If NULL, the function will return a summary across all
#'   generated simulation parameters. A character vector can also be included
#'   here, in which case the function will return summaries for only the
#'   selected parameter names. Finally, users can input a list, with the names
#'   of the elements corresponding to parameter names, and the values of each
#'   element providing a vector with the parameter values to be included in the
#'   summary. In all cases, the summary will collapse across all parameters or
#'   parameter values that are excluded from \code{params}.
#' @param func The function to be used to generate the summary (e.g., mean, sum).
#' @param ... Any other arguments to be passed to func.
#' @return Returns a data frame with one row per set of unique simulation tests.
#' @seealso \code{\link{simulate}}
#' @examples
#' # will calculate the mean for all 'sig' values in all simulations where N=200,
#' # and all simulations where N=300
#' summary(power_sim, 'sig', params=list(N=c(200, 300)))
#' @export
summary.simulation <- function(sim, outcome, params=NULL, func=mean, ...) {
    if (is.null(params)) {
        grid <- sim$tests
    } else if (is.list(params)) {
        paramNames <- names(params)
        grid <- sim$tests

        # it is assumed that each list element has a vector with the values to
        # be included for that parameter; so we loop through each parameter and
        # pare down the grid to only include those values
        for (p in 1:length(params)) {
            grid <- grid[grid[, paramNames[p]] %in% params[[p]], , drop=FALSE]
        }

        grid <- unique(grid[, paramNames, drop=FALSE])
    } else if (is.vector(params)) {
        grid <- unique(sim$tests[, params, drop=FALSE])
    }

    if (nrow(grid) > 0) {
        results <- grid
        # loop through each combination of tests
        for (i in 1:nrow(grid)) {
            data <- sim$results

            # loop through each parameter value individually and filter down data
            for (p in 1:ncol(grid)) {
                paramName <- paste0(names(sim$tests)[p], '.test')
                data <- data[data[, paramName] == sim$tests[i, p], ]
            }

            results[i, as.character(substitute(func))] <- func(data[, outcome], ...)
        }
    } else {
        results <- func(sim$results[, outcome], ...)
    }
    return(results)
}




R/gen_data.R
#' Generate data through a factor matrix and effects matrix.
#'
#' \code{gen_data} will generate sample data based on a factor structure and
#' effects structure specified by the user.
#'
#' @param factor_struct A matrix describing the measurement model of latent
#'   factors (columns) as measured by observed variables (rows).
#' @param effects_struct A matrix describing the variances and covariances of
#'   the latent variables in the model.
#' @param nCases Number of sample cases to generate.
#' @param trueScores Whether or not to include the data for each variable as
#'   measured without error. If set to TRUE, the resulting data frame will
#'   include all the variables in the model twice: once with measurement error,
#'   and once without.
#' @return Returns a data frame with \code{nCases} rows and columns for each
#'   observed and latent variable. These variables will approximately accord
#'   with the factor structure and effects structure that was specified, within
#'   sampling error.
#' @examples
#' # two uncorrelated predictors, one criterion, with measurement error in all
#' # variables
#' beta1 <- .5
#' beta2 <- .6
#' y_resid_var <- sqrt(1 - (beta1^2 + beta2^2))
#' fmodel <- matrix(
#'     c(.8, 0, 0,   # x1
#'       0, .6, 0,   # x2
#'       0, 0, .5),  # y
#'     nrow=3, ncol=3, byrow=TRUE, dimnames=list(
#'     c('x1', 'x2', 'y'), c('x1', 'x2', 'y')))
#'     # in this case, observed and latent variables are the same
#' effects <- matrix(
#'     c(1, 0, beta1,
#'       0, 1, beta2,
#'       0, 0, y_resid_var),
#'     nrow=3, ncol=3, byrow=TRUE, dimnames=list(
#'     c('x1', 'x2', 'y'), c('x1', 'x2', 'y')))
#'
#' sample_data <- gen_data(fmodel, effects, nCases=1000)
#' round(var(sample_data), 2)
#' round(cor(sample_data), 2)
#' summary(lm(y ~ x1 + x2, data=sample_data))
#'     # note that beta coefficients are much smaller, due to measurement error
#' @export
gen_data <- function(factor_struct, effects_struct, nCases=1000,
    trueScores=FALSE) {

    nVars <- dim(factor_struct)[1]  # problem size determined by input to the function
    nLatent <- dim(factor_struct)[2]

    tmodel <- t(factor_struct)  # transpose of model

    communality <- diag(factor_struct %*% tmodel)
        # find how much to weight true scores and errors given the measurement model
    uniqueness <- 1 - communality
    error_weight <- diag(sqrt(uniqueness))  # how much to weight the errors

    latent_scores <- matrix(rnorm(nCases * nLatent), nCases)
        # generate true scores for the latent variables
    latent_scores <- latent_scores %*% effects_struct
        # ensure true scores reflect structural relations between the factors

    true_scores <- latent_scores %*% tmodel
    error <- matrix(rnorm(nCases * nVars), nCases)
        # generate normal error
    wtd_error <- error %*% error_weight
    observed_scores <- true_scores + wtd_error

    if (trueScores) {
        colnames(true_scores) <- paste0(colnames(observed_scores), '_true')
        return(data.frame(observed_scores, true_scores))
    } else {
        return(data.frame(observed_scores))
    }
}







R/power_lm.R
# single variable linear regression
# gen_data <- function(N, vars=c('y', 'x'), coef=0, mean=0, sd=1) {
#     obs <- rep(1:N)
#
#     if (typeof(coef) == 'list') {
#         if ('b0' %in% names(coef)) {
#             b0 <- coef[['b0']]
#         } else {
#             b0 <- 0
#         }
#
#         if ('b1' %in% names(coef)) {
#             b1 <- coef[['b1']]
#         } else {
#             b1 <- 0
#         }
#     } else if (is.numeric(coef)) {
#         b0 <- coef
#         b1 <- coef
#     } else {
#         b0 <- 0
#         b1 <- 0
#     }
#
#     if (typeof(mean) == 'list') {
#         if ('x' %in% names(mean)) {
#             xm <- mean[['x']]
#         } else {
#             xm <- 0
#         }
#     } else if (is.numeric(mean)) {
#         xm <- mean
#     } else {
#         xm <- 0
#     }
#
#     if (typeof(sd) == 'list') {
#         if ('x' %in% names(sd)) {
#             xsd <- sd[['x']]
#         } else {
#             xsd <- 1
#         }
#     } else if (is.numeric(sd)) {
#         xsd <- sd
#     } else {
#         xsd <- 1
#     }
#
#     x <- rnorm(N, xm, xsd)
#     y <- rnorm(N, b0 + b1*x, (1-b1^2))
#     return(data.frame(obs, y, x))
# }
#
# power_lm1 <- function(N, vars=c('y', 'x'), form='y ~ x', coef=0, mean=0, sd=1,
#     n.sims=5000, seed=runif(1)*1000) {
#
#     formula <- formula(form)
#     allTerms <- character()
#     if (attr(terms(formula), 'intercept') == 1) {
#         allTerms <- 'intercept'
#     }
#     allTerms <- c(allTerms, attr(terms(formula), 'term.labels'))
#
#     dimnames <- 'sim'
#     for (v in 1:length(vars)) {
#         dimnames <- c(dimnames, paste0(vars[v], '.mean'), paste0(vars[v], '.sd'))
#     }
#
#     for (t in 1:length(allTerms)) {
#         dimnames <- c(dimnames, paste0(allTerms[t], '.est'),
#             paste0(allTerms[t], '.se'), paste0(allTerms[t], '.p'),
#             paste0(allTerms[t], '.sig'))
#     }
#
#     output <- matrix(nrow=n.sims, ncol=length(dimnames), dimnames=list(NULL,
#         dimnames))
#     set.seed(seed)
#     for (s in 1:n.sims) {
#         sim <- gen_data(N, vars, coef, mean, sd)
#         model <- lm(y ~ x, data=sim)
#
#         output[s, 'sim'] <- s
#         output[s, 'x mean'] <- mean(sim$x)
#         output[s, 'x sd'] <- sd(sim$x)
#         output[s, 'y mean'] <- mean(sim$y)
#         output[s, 'y sd'] <- sd(sim$y)
#         output[s, 'estimate'] <- coef(summary(model))['x', 'Estimate']
#         output[s, 'se'] <- coef(summary(model))['x', 'Std. Error']
#         output[s, 'p'] <- coef(summary(model))['x', 'Pr(>|t|)']
#         output[s, 'sig'] <- output[s, 'estimate'] > 0 & output[s, 'p'] <= .05
#     }
#     print(paste('Estimated power:', mean(output[, 'sig'])))
#
#     output <- as.data.frame(output)
#     output$sig <- as.logical(output$sig)
#     return(output)
# }
#
# system.time(power <- power_lm1(N=343, coef=list(b1=.15), n.sims=5000, seed=100))
# mean(power[['x mean']])
# mean(power[['x sd']])
# mean(power[['y mean']])
# mean(power[['y sd']])
# mean(power[['estimate']])
# mean(power[['se']])
# mean(power[['p']])
# mean(power[['sig']])








R/simulate.R
#' Simulate data and statistical models.
#'
#' \code{simulate} simulates data by generating samples according to a user-
#' defined function. These data can then be used to examine whatever statistic
#' you wish. This function is intentionally kept general and flexible, to allow
#' for a wide variety of applications.
#'
#' @param func A user-defined function that generates data and performs a
#'   statistical test on that data.
#' @param params A list of parameters to be passed to \code{func}. Each of these
#'   parameters can be a vector of multiple values, and \code{simulate} will
#'   test each combination of these values.
#' @param n.sims Number of simulations (per combination of params).
#' @param output Specifies how the function provides the ultimate results of the
#'   simulations: can return a "list", a "dataframe", or a "vector". Note that
#'   the output from the supplied function must be able to be coerced into this
#'   output type.
#' @param boot Whether or not to use bootstrapped data to pass along to your
#'   simulation.
#' @param bootParams If \code{boot=TRUE}, then use \code{bootParams} to pass
#'   along a named list of arguments to the \code{\link{boot}} function. The
#'   statistic and R parameters will be filled automatically, but at minimum you
#'   will need to pass along data. Information about parallel processing will
#'   also be passed along automatically.
#' @param parallel The type of parallel operation to be used (if any).
#' @param ncpus Integer: the number of processes to be used in parallel
#'   operation.
#' @param cl An optional \code{parallel} or \code{snow} cluster for use if
#'   \code{parallel = 'snow'}. If not supplied, a cluster on the local machine
#'   is created for the duration of the simulations.
#' @param beep Include a numeric value or character vector indicating the sound
#'   you wish to play once the simulation is done running. Requires the "beepr"
#'   package, and information about supported values is available in the
#'   documentation for that package.
#' @param ... Additional arguments to be passed to \code{func}. If you do not
#'   need to vary certain parameters in your model, you can pass them to
#'   \code{func} here.
#' @return Returns a list (by default) with one element per simulation. If
#'   \code{output} is specified as "dataframe", then \code{func} must
#'   return a (named) vector with the results you wish to capture; if
#'   \code{output} is specified as "vector", then \code{func} must return a
#'   one-element vector.
#' @seealso \code{\link{boot}}
#' @examples
#' lm_test <- function(N, b0, b1) {
#' x <- rnorm(N, 0, 1)
#' y <- rnorm(N, b0 + b1*x, sqrt(1 - b1^2))
#' data <- data.frame(y, x)
#' model <- lm(y ~ x, data)
#'
#' # capture output from model summary
#' est <- coef(summary(model))['x', 'Estimate']
#' se <- coef(summary(model))['x', 'Std. Error']
#' p <- coef(summary(model))['x', 'Pr(>|t|)']
#'
#' return(c(xm=mean(x), xsd=sd(x), ym=mean(y), ysd=sd(y), est=est, se=se, p=p,
#'     sig=est > 0 & p <= .05))
#' }
#'
#' # test power for two sample sizes: N=200 and N=300, with 5000 sims for each
#' power_sim <- simulate(lm_test, params=list(N=c(200, 300)), n.sims=5000, b0=0, b1=.15)
#' @export
simulate <- function(func, params=NULL, n.sims=5000,
    output=c('list', 'dataframe', 'vector'), boot=FALSE, bootParams=NULL,
    parallel=c('no', 'multicore', 'snow'), ncpus=1, cl=NULL, beep=NULL, ...) {

    dots <- list(...)
    outputType <- match.arg(output)

    # cross each param value with every other one, to create all combinations
    if (!is.null(params)) {
        grid <- expand.grid(params, KEEP.OUT.ATTRS=FALSE)
        grid_output <- grid
        names(grid_output) <- paste0(names(grid_output), '.test')
        nSets <- nrow(grid)
    } else {
        grid <- data.frame()  # empty
        grid_output <- data.frame()
        nSets <- 1
    }

    # figure out which parallel method to use
    parallel <- match.arg(parallel)
    have_mc <- FALSE
    have_snow <- FALSE
    if (parallel != 'no' && ncpus > 1) {
        if (requireNamespace('parallel', quietly=TRUE)) {
            if (parallel == 'multicore') {
                have_mc <- .Platform$OS.type != 'windows'
                if (!have_mc) {
                    warning('Multicore is not supported on Windows. Try parallel = "snow" instead. Proceeding with simulations serially.')
                }
            } else if (parallel == 'snow') {
                have_snow <- TRUE
            }
        } else {
            warning("Loading package 'parallel' failed. Proceeding with simulations serially.")
        }
        if (!have_mc && !have_snow) {
            ncpus <- 1
        }
    }

    # set up cluster if necessary
    if (ncpus > 1 && have_snow) {
        if (is.null(cl)) {
            clust <- parallel::makePSOCKcluster(rep('localhost',
                ncpus))
            if (RNGkind()[1] == "L'Ecuyer-CMRG") {
                parallel::clusterSetRNGStream(clust)
            }
        } else {
            clust <- cl
        }
    }

    cat(paste0('Running ',
        prettyNum(nSets * n.sims, big.mark=',', scientific=FALSE),
        ' simulations...\n'))

    allResults <- NULL  # variable to fill with final output

    timing <- system.time(
        for (set in 1:nSets) {
            # bootstrap data
            if (boot && 'data' %in% names(bootParams)) {
                boot_output <- do.call(boot::boot, args=c(list(statistic=func,
                    R=n.sims, parallel=parallel, ncpus=ncpus, cl=cl), bootParams,
                    c(as.list(grid[set, , drop=FALSE]), dots)))

                col_names <- names(boot_output$t0)
                output <- boot_output$t
                colnames(output) <- col_names

            # simulate data
            } else {
                if (ncpus > 1 &&
                    (have_mc || have_snow)) {

                    if (have_mc) {
                        output <- do.call(parallel::mclapply,
                            args=c(list(X=1:n.sims, FUN=func, mc.cores=ncpus),
                                c(as.list(grid[set, , drop=FALSE]), dots)))
                    } else if (have_snow) {
                        output <- do.call(parallel::parLapply,
                            args=c(list(cl=clust, X=1:n.sims, fun=func),
                                c(as.list(grid[set, , drop=FALSE]), dots)))
                    }
                } else {
                    output <- do.call(lapply, args=c(list(X=1:n.sims, FUN=func),
                        c(as.list(grid[set, , drop=FALSE]), dots)))
                }

                # convert output to data frame/vector if requested
                if (outputType == 'dataframe') {
                    col_names <- names(output[[1]])
                    output <- do.call(rbind.data.frame, output)
                    colnames(output) <- col_names
                } else if (outputType == 'vector') {
                    output <- unlist(output)
                }
            }

            if (outputType == 'dataframe') {
                rowsEachSim <- nrow(output) / n.sims
                if (!is.null(params)) {
                    extendGrid <- matrix(rep(unlist(grid_output[set, , drop=FALSE]),
                        each=n.sims), nrow=n.sims)
                    colnames(extendGrid) <- names(grid_output)

                    result <- cbind(sim=rep(1:n.sims, each=rowsEachSim), extendGrid, output)
                } else {
                    result <- cbind(sim=rep(1:n.sims, each=rowsEachSim), output)
                }

                if (is.null(allResults)) {
                    allResults <- result
                } else {
                    allResults <- rbind(allResults, result)
                }
                row.names(allResults) <- NULL
            } else {
                if (is.null(allResults)) {
                    allResults <- output
                } else {
                    allResults <- c(allResults, output)
                }
            }
        }
    )

    # stop cluster if we created it
    if (ncpus > 1 && have_snow && is.null(cl)) {
        parallel::stopCluster(clust)
    }

    if (outputType == 'dataframe') {
        allResults <- as.data.frame(allResults)
    }

    output <- list(results=allResults, tests=grid, n.sims=n.sims, timing=timing)
    class(output) <- 'simulation'

    # Ding! Fries are done
    if (!is.null(beep)) {
        if (requireNamespace('beepr', quietly=TRUE)) {
            beepr::beep(beep)
        }
    }

    return(output)
}




